src/software/thunderscope/gl/layers/gl_referee_info_layer.py

```diff
@@ -0,0 +1,269 @@
+
+        if not self.gamestate_type_text:
+            self.gamestate_type_text = GLLabel(
+                parent_item=self,
+                offset=(-10, 50),
+                text=GLRefereeInfoLayer.GAMESTATE_PREFIX + referee_msg_dict["stage"],
+            )
+            self.referee_text_graphics.append(self.gamestate_type_text)
+        else:
+            self.gamestate_type_text.set_text(
+                GLRefereeInfoLayer.GAMESTATE_PREFIX + referee_msg_dict["stage"]
+            )
+
+        if not self.command_type_text:
+            self.command_type_text = GLLabel(
+                parent_item=self,
+                offset=(-10, 70),
+                text=GLRefereeInfoLayer.GAMESTATE_PREFIX + referee_msg_dict["command"],
+            )
+            self.referee_text_graphics.append(self.command_type_text)
+        else:
+            self.command_type_text.set_text(
+                GLRefereeInfoLayer.REFEREE_COMMAND_PREFIX + referee_msg_dict["command"]
+            )
+
+    def refresh_graphics(self) -> None:
+        """Refresh all visuals for both ball placement and referee info"""
+        self.cached_world = self.world_buffer.get(block=False, return_cached=True)
+
+        self.__update_referee_info()
+        self.__update_ball_placement()
+
+    def __update_ball_placement_status(
+        self, ball_state: BallState, new_placement_point: Point
+    ) -> None:
+        """Update ball placement circle color corresponding to ball position.
+        If the ball lies inside the tolerance circle, the circle will be green, otherwise red.
+        :param ball_state: state of the ball
+        :param new_placement_point: new ball placement point to visualize
+        """
+        if GLRefereeInfoLayer.is_point_in_circle(
+            ball_state.global_position,
+            new_placement_point,
+            BALL_PLACEMENT_TOLERANCE_RADIUS_METERS,
+        ):
+            self.placement_tolerance_graphic.set_outline_color(
+                self.BALL_PLACEMENT_TOLERANCE_VISUALIZATION_COLOR_ON
+            )
+        else:
+            self.placement_tolerance_graphic.set_outline_color(
+                self.BALL_PLACEMENT_TOLERANCE_VISUALIZATION_COLOR_OFF
+            )
+
+    def __update_target_visual(self) -> None:
+        """Update the ball placement target graphic to shrink or expand"""
+        if self.shrink_target:
+            self.placement_target_graphic.set_radius(
+                self.placement_target_graphic.radius - 0.01
+            )
+            self.shrink_target = self.placement_target_graphic.radius >= 0
+        else:
+            self.placement_target_graphic.set_radius(
+                self.placement_target_graphic.radius + 0.01
+            )
+            self.shrink_target = (
+                self.placement_target_graphic.radius
+                >= BALL_PLACEMENT_TOLERANCE_RADIUS_METERS
+            )
+
+        # update the count-down graphics
+        time_left = max(
```

Comment:
this should be guarded with `if self.cached_referee_info`

Reply:
Done